File: contest-management-client-server-java/ContestMgmtPersistence/src/main/java/contestmgmt/persistence/repository/jdbc/CompetitionDbRepository.java
package contestmgmt.persistence.repository.jdbc;

import contestmgmt.model.Competition;
import contestmgmt.model.Tuple;
import contestmgmt.persistence.repository.ICompetitionRepository;

import java.sql.*;
import java.util.*;

public class CompetitionDbRepository implements ICompetitionRepository {
    private JdbcUtils dbUtils;

    public CompetitionDbRepository(Properties props) {
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Iterable<Competition> findAll() {
        Connection conn = dbUtils.getConnection();
        List<Competition> competitions = new ArrayList<>();

        try (PreparedStatement prepStmt = conn.prepareStatement("select * from Competitions")) {
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("id");
                    String competitionType = result.getString("competition_type");
                    String ageCategory = result.getString("age_category");
                    Competition comp = new Competition(competitionType, ageCategory);
                    comp.setId(id);
                    competitions.add(comp);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return competitions;
    }

    @Override
    public Competition find(Long aLong) {
        Connection conn = dbUtils.getConnection();
        Competition comp = null;
        try (PreparedStatement prepStmt = conn.prepareStatement("select * from Competitions where id = ?")) {
            prepStmt.setLong(1, aLong);
            try (ResultSet result = prepStmt.executeQuery()) {
                if (result.next()) {
                    long id = result.getLong("id");
                    String competitionType = result.getString("competition_type");
                    String ageCategory = result.getString("age_category");
                    comp = new Competition(competitionType, ageCategory);
                    comp.setId(id);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return comp;
    }

    @Override
    public Long add(Competition entity) {
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "insert into Competitions " +
                        "(competition_type, age_category) " +
                        "values (?, ?)", Statement.RETURN_GENERATED_KEYS)) {
            prepStmt.setString(1, entity.getCompetitionType());
            prepStmt.setString(2, entity.getAgeCategory());

            int result = prepStmt.executeUpdate();
            ResultSet resultSet = prepStmt.getGeneratedKeys();
            if (resultSet.next())
                return resultSet.getLong(1);
        } catch (SQLException e) {
            System.err.println("Error DB " + e);
        }
        return null;
    }

    @Override
    public void remove(Long aLong) {
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn.prepareStatement("delete from Competitions where id = ?")) {
            prepStmt.setLong(1, aLong);
            int result = prepStmt.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB " + e);
        }
    }

    @Override
    public void update(Competition entity) {
        Connection conn = dbUtils.getConnection();
        try (PreparedStatement prepStmt = conn
                .prepareStatement("update Competitions set competition_type = ?, age_category = ? where id = ?")) {
            prepStmt.setString(1, entity.getCompetitionType());
            prepStmt.setString(2, entity.getAgeCategory());
            prepStmt.setLong(3, entity.getId());
            int result = prepStmt.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB " + e);
        }
    }

    @Override
    public Map<Long, Tuple<Competition, Integer>> countParticipantsForEachCompetition(String competitionType,
                                                                                      String ageCategory) {
        Connection conn = dbUtils.getConnection();
        Map<Long, Tuple<Competition, Integer>> map = new HashMap<>();
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select *, count(competition_id) cnt " +
                        "from Competitions C " +
                        "left join Registrations R on C.id = R.competition_id " +
                        "where competition_type like '%' || ? || '%' and age_category like '%' || ? || '%' " +
                        "group by id, competition_id, competition_type, age_category")) {
            prepStmt.setString(1, competitionType);
            prepStmt.setString(2, ageCategory);
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("id");
                    String compType = result.getString("competition_type");
                    String ageCat = result.getString("age_category");
                    int count = result.getInt("cnt");
                    Competition c = new Competition(compType, ageCat);
                    c.setId(id);
                    map.put(id, new Tuple<>(c, count));
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB " + e);
        }
        return map;
    }

    @Override
    public Iterable<Competition> findCompetitionsByString(String competitionType, String ageCategory) {
        Connection conn = dbUtils.getConnection();
        List<Competition> competitions = new ArrayList<>();

        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select * from Competitions " +
                        "where competition_type like '%' || ? || '%' and age_category like '%' || ? || '%'")) {
            prepStmt.setString(1, competitionType);
            prepStmt.setString(2, ageCategory);
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    Long id = result.getLong("id");
                    String compType = result.getString("competition_type");
                    String ageCat = result.getString("age_category");
                    Competition c = new Competition(compType, ageCat);
                    c.setId(id);
                    competitions.add(c);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return competitions;
    }

    @Override
    public Iterable<String> findAgeCategoriesFromCompetitionType(String competitionType) {
        Connection conn = dbUtils.getConnection();
        List<String> ageCategories = new ArrayList<>();
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select age_category from Competitions " +
                        "where competition_type = ?")) {
            prepStmt.setString(1, competitionType);
            try (ResultSet result = prepStmt.executeQuery()) {
                while (result.next()) {
                    String ageCategory = result.getString("age_category");
                    ageCategories.add(ageCategory);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return ageCategories;
    }

    @Override
    public Competition findByProps(String competitionType, String ageCategory) {
        Connection conn = dbUtils.getConnection();
        Competition c = null;
        try (PreparedStatement prepStmt = conn.prepareStatement(
                "select * from Competitions " +
                        "where competition_type = ? and age_category = ?")) {
            prepStmt.setString(1, competitionType);
            prepStmt.setString(2, ageCategory);
            try (ResultSet result = prepStmt.executeQuery()) {
                if (result.next()) {
                    Long id = result.getLong("id");
                    String compType = result.getString("competition_type");
                    String ageCat = result.getString("age_category");
                    c = new Competition(compType, ageCat);
                    c.setId(id);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error DB" + e);
        }
        return c;
    }
}

File: contest-management-client-server-cs/persistence/database/CompetitionDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Data.SQLite;
using model;
using persistence.utils;

namespace persistence.database
{
    public class CompetitionDbRepository : ICompetitionRepository
    {
        public IEnumerable<Competition> FindAll()
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();
            
            using var cmd = conn.CreateCommand();
            cmd.CommandText = @"select * from Competitions";
            try
            {
                using var reader = cmd.ExecuteReader();
                var competitions = new List<Competition>();
                while (reader.Read())
                {
                    competitions.Add(new Competition
                    {
                        Id = reader.GetInt64(0),
                        CompetitionType = reader.GetString(1),
                        AgeCategory = reader.GetString(2)
                    });
                }

                return competitions;
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB" + e);
            }

            return new List<Competition>();
        }

        public Competition? Find(long id)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();
            
            using var cmd = conn.CreateCommand();
            cmd.CommandText = 
                @"select * from Competitions
                 where id = @id";
            cmd.AddParameterWithValue("@id", id);
            try
            {
                using var reader = cmd.ExecuteReader();
                if (reader.Read())
                    return new Competition
                    {
                        Id = reader.GetInt64(0),
                        CompetitionType = reader.GetString(1),
                        AgeCategory = reader.GetString(2),
                    };
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB" + e);
            }
            
            return default;
        }

        public void Add(Competition entity)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = 
                @"insert into Competitions (competition_type, age_category) VALUES
                (@competition_type, @age_category)";
            cmd.AddParameterWithValue("@competition_type", entity.CompetitionType);
            cmd.AddParameterWithValue("@age_category", entity.AgeCategory);
            try
            {
                entity.Id = (long) (cmd.ExecuteScalar() ?? throw new NullReferenceException());
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB " + e);
            }
        }

        public void Remove(long id)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "delete from Competitions where id = @id";
            cmd.AddParameterWithValue("@id", id);
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB " + e);
            }
        }

        public void Update(Competition entity)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText =
                @"update Competitions set
                competition_type = @competition_type,
                age_category = @age_category
                where id = @id";
            cmd.AddParameterWithValue("@competition_type", entity.CompetitionType);
            cmd.AddParameterWithValue("@age_category", entity.AgeCategory);
            cmd.AddParameterWithValue("@id", entity.Id);
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB " + e);
            }
        }

        public IDictionary<long, (Competition, int)> GetCompetitionsAndCounts(string competitionTypeStr,
            string ageCategoryStr)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText =
                @"select id, competition_type, age_category, count(competition_id) cnt from Competitions C
                left join Registrations R on C.id = R.competition_id
                where
                competition_type like '%' || @competition_type || '%' and
                age_category like '%' || @age_category || '%'
                group by id, competition_id, competition_type, age_category";
            cmd.AddParameterWithValue("@competition_type", competitionTypeStr);
            cmd.AddParameterWithValue("@age_category", ageCategoryStr);
            try
            {
                using var reader = cmd.ExecuteReader();
                var dict = new Dictionary<long, (Competition, int)>();
                while (reader.Read())
                {
                    var id = reader.GetInt64(0);
                    dict.Add(id,
                    (
                        new Competition
                        {
                            Id = id,
                            CompetitionType = reader.GetString(1),
                            AgeCategory = reader.GetString(2)
                        },
                        reader.GetInt32(3)
                    ));
                }

                return dict;
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB " + e);
            }

            return new Dictionary<long, (Competition, int)>();
        }

        public IEnumerable<string> FindCompetitionTypes(string competitionTypeStr)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText =
                @"select distinct competition_type from Competitions
                where
                competition_type like '%' || @competition_type || '%'";
            cmd.AddParameterWithValue("@competition_type", competitionTypeStr);
            try
            {
                using var reader = cmd.ExecuteReader();
                var competitionTypes = new List<string>();
                while (reader.Read())
                {
                    competitionTypes.Add(reader.GetString(0));
                }

                return competitionTypes;
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB" + e);
            }

            return new List<string>();
        }

        public IEnumerable<string> FindAgeCategoriesFromCompetitionType(string competitionType)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText =
                @"select age_category from Competitions
                where competition_type = @competition_type";
            cmd.AddParameterWithValue("@competition_type", competitionType);
            try
            {
                using var reader = cmd.ExecuteReader();
                var ageCategories = new List<string>();
                while (reader.Read())
                    ageCategories.Add(reader.GetString(0));
                return ageCategories;
            }
            catch (DbException e)
            {
                Console.WriteLine("Error DB " + e);
            }

            return new List<string>();
        }

        public Competition? FindByProps(string competitionType, string ageCategory)
        {
            using var conn = DbUtils.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText =
                @"select * from Competitions
                where competition_type = @competition_type and age_category = @age_category";
            cmd.AddParameterWithValue("@competition_type", competitionType);
            cmd.AddParameterWithValue("@age_category", ageCategory);
            try
            {
                using var reader = cmd.ExecuteReader();
                if (reader.Read())
                    return new Competition
                    {
                        Id = reader.GetInt64(0),
                        CompetitionType = reader.GetString(1),
                        AgeCategory = reader.GetString(2)
                    };
            }
            catch (SQLiteException e)
            {
                Console.WriteLine("Error DB" + e);
            }

            return default;
        }
    }
}